core/unit_test/hip/TestHIP_AsyncLauncher.cpp

```diff
@@ -0,0 +1,95 @@
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <hip/TestHIP_Category.hpp>
+
+namespace Test {
+
+struct TestAsyncLauncher {
+  ;
+  volatile size_t *m_flag;
+  size_t m_value;
+
+  struct TagTest {};
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const TagTest &, const int i) const {
+    // and update flag
+    Kokkos::atomic_add(m_flag, m_value);
+  }
+
+  TestAsyncLauncher(volatile size_t *flag, int value)
+      : m_flag(flag), m_value(value) {}
+
+  void run() {
+    Kokkos::parallel_for(Kokkos::RangePolicy<TagTest>(0, 1), *this);
+  }
+};
+
+TEST(hip, async_launcher) {
+  size_t *flag;
+  HIP_SAFE_CALL(hipMalloc(&flag, sizeof(size_t)));
+  HIP_SAFE_CALL(hipMemset(flag, 0, sizeof(size_t)));
+  // launch # of cycles * 1000 kernels w/ distinct values
+  auto space        = Kokkos::Experimental::HIP();
+  auto instance     = space.impl_internal_space_instance();
+  size_t max_cycles = instance->m_maxDriverCycles;
+  size_t nkernels   = max_cycles * 1000;
+  std::vector<TestAsyncLauncher> launchers;
+  for (size_t i = 0; i < nkernels; ++i) {
+    launchers.push_back(TestAsyncLauncher(flag, i));
+    launchers[i].run();
```

Comment:
Why do you use a vector at all?
```suggestion
  std::vector<TestAsyncLauncher> launchers;
  for (size_t i = 0; i < nkernels; ++i) {
    TestAsyncLauncher(flag, i).run();
```

Reply:
Good question -- I think it's a relic from an older version of the test.

Reply:
switched to the suggested.